### DISCUSSION TITLE
Broke due to 'RecursionError: maximum recursion depth exceeded while calling a Python object' (and a temporary fix) (FIXED)

### DISCUSSION THREAD

**Discussion by dpmiller94** [DrewThomasson/VoxNovel#10] [General]
So, when I would try to run the program, I would get this error whenever trying to generate audio for a particularly long sentence:

<pre>Exception in thread Thread-1 (generate_audio):
Traceback (most recent call last):
  File &quot;/home/david/miniconda/envs/VoxNovel/lib/python3.10/threading.py&quot;, line 1016, in _bootstrap_inner
    self.run()
  File &quot;/home/david/miniconda/envs/VoxNovel/lib/python3.10/threading.py&quot;, line 953, in run
    self._target(*self._args, **self._kwargs)
  File &quot;/home/david/VoxNovel/gui_run.py&quot;, line 1835, in generate_audio
    fragments = split_long_sentence(sentence)
  File &quot;/home/david/VoxNovel/gui_run.py&quot;, line 1226, in split_long_sentence
    return split_long_sentence(first_half, max_length, max_pauses) + split_long_sentence(second_half, max_length, max_pauses)
  File &quot;/home/david/VoxNovel/gui_run.py&quot;, line 1226, in split_long_sentence
    return split_long_sentence(first_half, max_length, max_pauses) + split_long_sentence(second_half, max_length, max_pauses)
  File &quot;/home/david/VoxNovel/gui_run.py&quot;, line 1226, in split_long_sentence
    return split_long_sentence(first_half, max_length, max_pauses) + split_long_sentence(second_half, max_length, max_pauses)
  [Previous line repeated 990 more times]
  File &quot;/home/david/VoxNovel/gui_run.py&quot;, line 1216, in split_long_sentence
    closest_split = min(possible_splits, key=lambda x: abs(x - middle_index))
  File &quot;/home/david/VoxNovel/gui_run.py&quot;, line 1216, in &lt;lambda&gt;
    closest_split = min(possible_splits, key=lambda x: abs(x - middle_index))
RecursionError: maximum recursion depth exceeded while calling a Python object
</pre>

So it seems like something was going wrong with the 'split_long_sentence' function. I went and changed the code in gui_run.py from this:

```
def split_long_sentence(sentence, max_length=250, max_pauses=10):
    Recursively splits a sentence based on length or number of pauses.
    
    :param sentence: The sentence to split.
    :param max_length: Maximum allowed length of a sentence.
    :param max_pauses: Maximum allowed number of pauses in a sentence.
    :return: A list of sentence parts that meet the criteria.
    """
    # Check if the sentence meets the splitting criteria
    if len(sentence) >= max_length or sentence.count(',') + sentence.count(';') + sentence.count('.') > max_pauses:
        # Find the best place to split the sentence (middle pause or just the middle)
        possible_splits = [i for i, char in enumerate(sentence) if char in ',;.']
        
        if possible_splits:
            # Find the closest split point to the middle
            middle_index = len(sentence) // 2
            closest_split = min(possible_splits, key=lambda x: abs(x - middle_index))
        else:
            # If no punctuation to split on, choose the middle of the sentence
            closest_split = len(sentence) // 2
        
        # Split the sentence
        first_half = sentence[:closest_split + 1].strip()
        second_half = sentence[closest_split + 1:].strip()
        
        # Recursively split each half if necessary
        return split_long_sentence(first_half, max_length, max_pauses) + split_long_sentence(second_half, max_length, max_pauses)
    else:
        # If the sentence doesn't need splitting, return it as a single element list
        return [sentence]
```
        
to this:
        
```
def split_long_sentence(sentence, max_length=250, max_pauses=10, depth_limit=100):
    if depth_limit <= 0 or (len(sentence) < max_length and sentence.count(',') + sentence.count(';') + sentence.count('.') <= max_pauses):
        return [sentence]
    
    possible_splits = [i for i, char in enumerate(sentence) if char in ',;.']
    if not possible_splits:
        if len(sentence) > max_length:
            return [sentence[:max_length]] + split_long_sentence(sentence[max_length:], max_length, max_pauses, depth_limit - 1)
        else:
            return [sentence]
    
    middle_index = len(sentence) // 2
    closest_split = min(possible_splits, key=lambda x: abs(x - middle_index))
    
    first_half = sentence[:closest_split + 1].strip()
    second_half = sentence[closest_split + 1:].strip()
    
    parts = []
    if first_half:
        parts += split_long_sentence(first_half, max_length, max_pauses, depth_limit - 1)
    if second_half:
        parts += split_long_sentence(second_half, max_length, max_pauses, depth_limit - 1)
    
    return parts
```

   
And it seems to work now. In the terminal, I will still occasionally get this warning:
    
[!] Warning: The text length exceeds the character limit of 250 for language 'en', this might cause truncated audio.

But it doesn't seem to negatively impact the audio to any meaningful way. This fix was a lot of trial and error because, if it isn't already clear, I barely know what I'm doing, but I am sure the author would be able to see what was going wrong and patch it pretty easily.

**Comment by DrewThomasson** [comment]
OMG thx for the very informative bug report!!!

I actually made a non-recursive version of the sentence splitter function just in case this was possible, I should be able to apply it tomorrow.


This is rlly helpful and good to know tho!



Extra info:

The issue seems to be stemming from that I made it recursive thinking it would speed up the function but, nah this isn't worth it lol 

With the 250 limit thing yeah that means it might cut off the audio at the tts limit with your implementation if it's too long for the Xtts 'en' model to use

**Reply by dpmiller94** [reply]
I'm happy to help, its been a great program! Also, if you're interested in any potential feature requests, I've been using VoxNovel for non-fiction audiobooks and there are a few unique aspects of nonfiction formatting that require some manual tinkerings, specifically issues surrounding quotations and references. 

Since non-fiction is generally read in the second person, 'quotes' are typically not a part of dialogue. Instead, it is usually just done to reference a specific thing that someone/something said. However, in program, it recognizes quotes as being first-person character dialogue and tries to assign it to a specific 'character' (which can result in funny results, such as having 'chatacters' called "the newspaper" or "the report"). The way I fixed this is by just going through my epubs and replacing all quotation marks with backticks (`), which aren't flagged as character dialogue. Its worked on the three books I've tried so far, but I'm not sure if this alters anything with the TTS, such as making it so the reader is less likely to use intonations saying someone else's quotations. 

The second one is a bit more complex. In-text citations usually come in two formats: a number that links to a footnote or by citing the author/year/page# in parenthesis, such as (Collins and Miller, 2005, pg. 34). To prevent VoxNovel from saying these citations out-loud, I have been experimenting with creating a program that will automatically detect and delete these using R (which I know is inferior to python, but my coding knowledge comes from stats classes, unfortunately). I will basically do this:

1. Use calibre to extract each chapter of the book I want to fix
2. Load the corresponding HTML file for said chapter into R
3. Run the code that can detect and delete the footnotes or the in-text citations
4. Put the chapter(s) back into the epub (or just run each chapter through VoxNovel individually). 

I'm not sure if it would be possible to have a way to automate this in VoxNovel. I was thinking something like box you can tick for non-fiction that, if selected, will recognize that quotes are still done by the narrator and try to detect and delete in-text citations. If this sounds like something you are interested in and would like to see what I was doing, let me know! I know I'm not good enough at this to make these alterations on my own but I'd be happy to help if you thought it was a good feature to add!

**Comment by DrewThomasson** [comment]
I applied my non recursive split_long_sentence function to the main branch, 

you should just be able to do a git pull then, 

Should fix any issues you were having with it before, worked on my end with my basic testing file, 

let me know if you run into any other issue or if your problem still persists!


ALSO:
hmmm your suggestions look promising, I'll review them further for a more in depth response in a bit,


also ill look more into your solution you made for the function cause you might be able to fix your warning:

 "[!] Warning: The text length exceeds the character limit of 250 for language 'en', this might cause truncated audio."   

by lowering the max_length value to 249 or 230 instead of 250

**Comment by DrewThomasson** [comment]
Update on my repose to your comment:

Are you saying that for non-fiction books you usually try to make it so that everything in the book is only said by one voice actor?
-If so there is a option in the gui where it'll make everything said by the narrators voice actor.


<img width="755" alt="Screenshot 2024-03-18 at 8 25 34 PM" src="https://github.com/DrewThomasson/VoxNovel/assets/126999465/a720fa40-7631-4d85-81a1-aa93afd9571a">

-if not I may by confused lol













👀

For the second part also, I'd LOVE to see any code you've written for it even if its in R,
 I should be able to convert any code you've written from R top python

It sounds like a really promising feature to add to the program! 
If we got it working I could add a button in the gui to turn on that function.
